src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -1,197 +1,599 @@
+                ThrowMinMaxValueSwapped();
             }
 
-            long range = (long)maxValue - minValue;
-            return range <= int.MaxValue ?
-                (int)(Sample() * range) + minValue :
-                (int)((long)(GetSampleForLargeRange() * range) + minValue);
+            Debug.Assert(minValue == maxValue);
+            return minValue;
         }
 
-        public virtual double NextDouble() => Sample();
+        /// <summary>Returns a random floating-point number that is greater than or equal to 0.0, and less than 1.0.</summary>
+        /// <returns>A single-precision floating point number that is greater than or equal to 0.0, and less than 1.0.</returns>
+        public virtual float NextSingle() => _impl.NextSingle();
 
+        /// <summary>Returns a random floating-point number that is greater than or equal to 0.0, and less than 1.0.</summary>
+        /// <returns>A double-precision floating point number that is greater than or equal to 0.0, and less than 1.0.</returns>
+        public virtual double NextDouble() => _impl.NextDouble();
+
+        /// <summary>Fills the elements of a specified array of bytes with random numbers.</summary>
+        /// <param name="buffer">The array to be filled with random numbers.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="buffer"/> is null.</exception>
         public virtual void NextBytes(byte[] buffer)
         {
             if (buffer is null)
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.buffer);
             }
 
-            for (int i = 0; i < buffer.Length; i++)
-            {
-                buffer[i] = (byte)InternalSample();
-            }
+            _impl.NextBytes(buffer);
         }
 
-        public virtual void NextBytes(Span<byte> buffer)
+        /// <summary>Fills the elements of a specified span of bytes with random numbers.</summary>
+        /// <param name="buffer">The array to be filled with random numbers.</param>
+        public virtual void NextBytes(Span<byte> buffer) => _impl.NextBytes(buffer);
+
+        /// <summary>Returns a random floating-point number between 0.0 and 1.0.</summary>
+        /// <returns>A double-precision floating point number that is greater than or equal to 0.0, and less than 1.0.</returns>
+        protected virtual double Sample() => _impl.Sample();
+
+        private static void ThrowMinMaxValueSwapped() =>
+            throw new ArgumentOutOfRangeException("minValue", SR.Format(SR.Argument_MinMaxValue, "minValue", "maxValue"));
+
+        private static void ThrowMaxValueMustBeNonNegative() =>
+            throw new ArgumentOutOfRangeException("maxValue", SR.Format(SR.ArgumentOutOfRange_NeedNonNegNum, "maxValue"));
+
+        /// <summary>Base type for all generator implementations that plug into the base Random.</summary>
+        private abstract class ImplBase
```

Comment:
Why not interface? Any opportunity to share some code?

Reply:
Generally interface dispatch is slower than virtual dispatch, and there's no benefit here to using an interface.